server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -716,11 +716,13 @@ public LuceneDocument doc() {
                 return document;
             }
         };
-        // Disable tracking array scopes for ignored source, as it would be added to the parent doc.
-        // Nested documents are added to preserve object structure within arrays of objects, so the use
-        // of ignored source for arrays inside them should be mostly redundant.
+        // Since nested objects are separate documents, object structure above them
+        // is not relevant for purposes of storing source values for synthetic source.
+        // Synthetic source is generated by a unique synthetic field loader for
+        // every nested document meaning that there are no conflicts between different objects in an array
+        // that are possible for plain objects and that need to be handles by a second parsing pass.
+        // It is however possible that inside this nested object we'll eventually be in the array scope.
         cloned.inArrayScope = false;
```

Comment:
Previously we completely disabled second parsing pass logic for nested objects. However since `NameValue` contains current lucene doc it actually works as expected and ignored values are added to correct document.